vslib/src/sai_vs_switch.cpp

```diff
@@ -1,5 +1,287 @@
+                {
+                    SWSS_LOG_DEBUG("port oper status didn't changed, will not send notification");
+                    return;
+                }
+            }
+
+            update_port_oper_status(port_id, data.port_state);
+
+            SWSS_LOG_DEBUG("executing callback SAI_SWITCH_ATTR_PORT_STATE_CHANGE_NOTIFY for port %s: %s",
+                    sai_serialize_object_id(data.port_id).c_str(),
+                    sai_serialize_port_oper_status(data.port_state).c_str());
+
+            callback(1, &data);
+        }
+
+    private:
+
+        sai_object_id_t m_switch_id;
+};
+
+void linkmsg_thread(
+        _In_ sai_object_id_t switch_id)
+{
+    SWSS_LOG_ENTER();
+
+    std::shared_ptr<SwitchState> sw = vs_get_switch_state(switch_id);
+
+    if (sw == nullptr)
+    {
+        SWSS_LOG_ERROR("failed to get switch state for switch id %s",
+                sai_serialize_object_id(switch_id).c_str());
+        return;
+    }
+
+    LinkMsg linkMsg(switch_id);
+
+    swss::NetDispatcher::getInstance().registerMessageHandler(RTM_NEWLINK, &linkMsg);
+    swss::NetDispatcher::getInstance().registerMessageHandler(RTM_DELLINK, &linkMsg);
+
+    SWSS_LOG_NOTICE("netlink msg listener started for switch %s",
+            sai_serialize_object_id(switch_id).c_str());
+
+    while (sw->getRunLinkThread())
+    {
+        try
+        {
+            swss::NetLink netlink;
+
+            swss::Select s;
+
+            netlink.registerGroup(RTNLGRP_LINK);
+            netlink.dumpRequest(RTM_GETLINK);
+
+            s.addSelectable(&netlink);
+            s.addSelectable(sw->getLinkThreadEvent());
+
+            while (sw->getRunLinkThread())
+            {
+                swss::Selectable *temps = NULL;
+
+                int tempfd;
+                int result = s.select(&temps, &tempfd);
+
+                SWSS_LOG_INFO("select ended: %d", result);
+            }
+        }
+        catch (const std::exception& e)
+        {
+            SWSS_LOG_ERROR("exception: %s", e.what());
+            return;
+        }
+    }
+}
+
+void vs_create_netlink_message_listener(
+        _In_ sai_object_id_t switch_id)
+{
+    SWSS_LOG_ENTER();
+
+    if (g_vs_hostif_use_tap_device == false)
+    {
+        return;
+    }
+
+    auto sw = vs_get_switch_state(switch_id);
+
+    if (sw == nullptr)
+    {
+        SWSS_LOG_ERROR("failed to get switch state for switch id %s",
+                sai_serialize_object_id(switch_id).c_str());
+        return;
+    }
+
+    sw->setRunLinkThread(true);
+
+    std::shared_ptr<std::thread> linkThread =
+        std::make_shared<std::thread>(linkmsg_thread, switch_id);
+
+    linkThread->detach();
+
+    sw->setLinkThread(linkThread);
+}
+
+void vs_remove_netlink_message_listener(
+        _In_ sai_object_id_t switch_id)
+{
+    SWSS_LOG_ENTER();
+
+    if (g_vs_hostif_use_tap_device == false)
+    {
+        return;
+    }
+
+    auto sw = vs_get_switch_state(switch_id);
```

Comment:
what about initial value of port status? should we do a query for all vports and get their initial up/down status when we create our switch?